File: 3623-CountNumberofTrapezoidsI/3623-CountNumberofTrapezoidsI.py
# Last updated: 20/7/2025, 10:01:11 am
mod = 10**9 + 7

@cache
def fact(x):
    return factorial(x)

class Solution:
    def countTrapezoids(self, points: List[List[int]]) -> int:
        n = len(points)

        hor = defaultdict(int)

        for x, y in points:
            hor[y] += 1

        if len(hor) == 1: return 0
        ret = 0
        prev = 0
        # print(hor)
        for h in hor:
            cnt = hor[h]
            if cnt == 1: continue
            curr = cnt * (cnt - 1) // 2
            ret += prev * curr % mod
            ret %= mod
            prev += curr
            prev %= mod

        return ret

File: 3623-CountNumberofTrapezoidsI/3623-CountNumberofTrapezoidsI.cpp
// Last updated: 2/12/2025, 5:46:06 pm
1class Solution {
2public:
3    int countTrapezoids(vector<vector<int>>& points) {
4        int mod = 1e9 + 7;
5        unordered_map<int, int> slopes;
6
7        for (auto& cord: points) {
8            slopes[cord[1]]++;
9        }
10
11        int ans = 0;
12        int prev = 0;
13        for (auto& it: slopes) {
14            long long curr = it.second * 1LL * (it.second - 1) / 2;
15            ans = (ans + prev * curr % mod) % mod;
16            prev += curr;
17        }
18
19        return ans;
20    }
21};